api.js

```diff
@@ -720,6 +725,458 @@ authenticate.clientCache  = clientCache;
+        }, function(clientId, callback) {
+          var ext = undefined;
+
+          // Parse authorization header for ext
+          var attrs = hawk.utils.parseAuthorizationHeader(
+            req.authorization
+          );
+          // Extra ext
+          if (!(attrs instanceof Error)) {
+            ext = attrs.ext;
+          }
+
+          // Get credentials with ext
+          loadCredentials(clientId, ext, callback);
+        }, {
+          // Not sure if JSON stringify is not deterministic by specification.
+          // I suspect not, so we'll postpone this till we're sure we want to do
+          // payload validation and how we want to do it.
+          //payload:      JSON.stringify(req.body),
+
+          // We found that clients often have time skew (particularly on OSX)
+          // since all our services require https we hardcode the allowed skew
+          // to a very high number (15 min) similar to AWS.
+          timestampSkewSec: 15 * 60,
+
+          // Provide nonce manager
+          nonceFunc:    options.nonceManager
+        }, authenticated);
+      } else {
+      // If there is no authorization header we'll attempt a login with bewit
+        hawk.uri.authenticate({
+          method:           req.method.toUpperCase(),
+          url:              req.resource,
+          host:             req.host,
+          port:             req.port,
+          authorization:    req.authorization
+        }, function(clientId, callback) {
+          var ext = undefined;
+
+          // Get bewit string (stolen from hawk)
+          var parts = req.resource.match(
+            /^(\/.*)([\?&])bewit\=([^&$]*)(?:&(.+))?$/
+          );
+          var bewitString = hoek.base64urlDecode(parts[3]);
+          if (!(bewitString instanceof Error)) {
+            // Split string as hawk does it
+            var parts = bewitString.split('\\');
+            if (parts.length === 4 && parts[3]) {
+              ext = parts[3];
+            }
+          }
+
+          // Get credentials with ext
+          loadCredentials(clientId, ext, callback);
+        }, {}, authenticated);
+      }
+    });
+  };
+};
+
+/**
+ * Make a function for the remote signature validation.
+ *
+ * options:
+ * {
+ *    authBaseUrl:   'https://....' // baseUrl for authentication server
+ * }
+ *
+ * The function returns takes an object:
+ *     {method, url, host, port, authorization}
+ * And return promise for an object on one of the forms:
+ *     {message}, {scheme, scopes}, {schema, scopes, hash}
+ *
+ * The method returned by this function works as `signatureValidator` for
+ * `remoteAuthentication`.
+ */
+var createRemoteSignatureValidator = function(options) {
+  assert(options.authBaseUrl, "options.authBaseUrl is required");
+  var agent = null;
+  if (/^https:/.test(options.authBaseUrl)) {
+    agent = new https.Agent({
+      keepAlive:  true,
+      maxSockets: 50
+    });
+  } else {
+    agent = new http.Agent({
```

Comment:
Yeah, for testing we use a local auth server... There is a mini one included in the `testing.js`.
Anyways, we should use taskcluster client for this later... so we have the same retry logic as elsewhere for free...

So consider this a temp hack until it's rolled out for `auth.taskcluster.net`.